text/0000-combined-restorer-analyzer-phases.md

```diff
@@ -0,0 +1,130 @@
+# Meta
+[meta]: #meta
+- Name: Combined restorer/analyzer and cacher/exporter phases.
+- Start Date: 2019-08-02
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Though not formally part of the spec, the project description and the
+lifecycle reference implementation contain the following four phases:
+
+- `/lifecycle/restorer` (considered an optional platform step)
+- `/lifecycle/analyzer`
+- `/lifecycle/exporter`
+- `/lifecycle/cacher` (considered an optional platform step)
+
+This proposal seeks to combine `restorer` with `analyzer` (to be named
+`restorer`) and `cacher` with `exporter` (to be named `exporter`).
+
+# Motivation
+[motivation]: #motivation
+
+In the current lifecycle reference implementation, the `restorer` phase
+fetches cached layers from a previous cache image. These layers can be
+arbitrarily large and, when using a remote repository, can represent a large
+amount of overhead. The `analyzer` phase runs after `restorer` and it
+may deterine that particular cache layers are no longer valid and remove them
+from the rest of the process - however, we've already paid a large cost to
+fetch them in the first place.
+
+The determination of cache layer validity can frequently be performed
+through the analysis of metadata: comparing metadata from the cache image with
+metadata from the previous application image. However, this comparison can
+only be performed with a combined `restorer` and `analyzer` phase.
+
+A parallel optimization potential exists when combining the `cacher` and
+`exporter` phases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Combined `restorer` phase.
+
+The spec will formalize the (combined) `restorer` phase like this:
+
+```
+/lifecycle/restorer \
+  -group=[(required) group.toml file] \
```

Comment:
Currently neither `-group` nor `-layers` are required. I think we should leave them as optional and keep the current defaulting behavior (first to an env var if it is present, then to a hardcoded value).

Reply:
I wasn't intending on removing the "env var overridable by flag" behavior of the current lifecycle implementation. I will add a section to formalize it.

But, even with env var, are `-group` and `-layers` _completely_ optional? E.g., is it valid to have no `CNB_LAYERS_DIR` and no `-layers` flag?

Reply:
When there is no env var the lifecycle falls back to a hardcoded default path https://github.com/buildpack/lifecycle/blob/master/cmd/cmd.go#L13